src/garage/tf/algos/npo.py

```diff
@@ -192,6 +203,87 @@ def optimize_policy(self, itr, samples_data):
         pol_ent = self._f_policy_entropy(*policy_opt_input_values)
         tabular.record('{}/Entropy'.format(self.policy.name), np.mean(pol_ent))
 
+        if self._fit_baseline_order == 'after':
+            self._fit_baseline_after(samples_data)
+
+        ev = np_tensor_utils.explained_variance_1d(samples_data['baselines'],
+                                                   samples_data['returns'],
+                                                   samples_data['valids'])
+        tabular.record('{}/ExplainedVariance'.format(self.baseline.name), ev)
+
+    def _fit_baseline(self, samples_data):
+        """Update baselines from samples.
+
+        Args:
+            samples_data (dict): Processed sample data.
+                See process_samples() for details.
+
+        """
+        policy_opt_input_values = self._policy_opt_input_values(samples_data)
+
+        # Augment reward from baselines
+        rewards_tensor = self._f_rewards(*policy_opt_input_values)
+        returns_tensor = self._f_returns(*policy_opt_input_values)
+        returns_tensor = np.squeeze(returns_tensor, -1)
+
+        paths = samples_data['paths']
+        valids = samples_data['valids']
+
+        # Recompute parts of samples_data
+        aug_rewards = []
+        aug_returns = []
+        for rew, ret, val, path in zip(rewards_tensor, returns_tensor, valids,
+                                       paths):
+            path['rewards'] = rew[val.astype(np.bool)]
+            path['returns'] = ret[val.astype(np.bool)]
+            aug_rewards.append(path['rewards'])
+            aug_returns.append(path['returns'])
+
+        # dense form
+        samples_data['rewards'] = np_tensor_utils.pad_tensor_n(
+            aug_rewards, self.max_path_length)
+        samples_data['returns'] = np_tensor_utils.pad_tensor_n(
+            aug_returns, self.max_path_length)
+
+        # Fit baseline
+        logger.log('Fitting baseline...')
+        self.baseline.fit(paths)
+
+    def _fit_baseline_before(self, samples_data):
+        """Update baselines from samples before optimization.
+
+        This function can be overrided by child class to perform
+        specific logic, e.g. if an algorithm fits baseline
+        after prediction, this function should be overrided and pass
+        only.
+
+        Args:
+            samples_data (dict): Processed sample data.
+                See process_samples() for details.
+
+        """
+        self._fit_baseline(samples_data)
+
+        # update baseline prediction for optimization
+        paths = samples_data['paths']
+        baselines = [self.baseline.predict(path) for path in paths]
```

Comment:
this is extremely strange and hard to follow. your method name says you are fitting the baseline, but then you also run the prediction step here. that's impossible to debug.

it doesn't seem like you even need different versions of `fit_baseline_before` and `fit_baseline_after` -- they both do the same thing. 

you should represent this in the logic of optimizer_policy instead. hiding it behind these methods is very confusing.